### PR TITLE
Python3 compatibility fixes

### PR DESCRIPTION
**Pull request by jdeschenes**
This is a fix to remove 2to3 when installing enable.

### CODE REVIEW

**Comment by jdeschenes** [comment]
The failure doesn't have anything to do with the actual PR. Should probably resolved separately.

**Comment by jdeschenes** [comment]
The travis-ci failure has been resolvedin #285.

File: enable/base.py [hunk 1]
```diff
@@ -103,8 +104,8 @@ def str_to_font ( object, name, value ):
                     underline = underline)
     except:
         pass
-    raise TraitError, ( object, name, 'a font descriptor string',
-                        repr( value ) )
+    raise TraitError(( object, name, 'a font descriptor string',
+                        repr( value ) ))
```
Comment (jwiggins) [inline enable/base.py]: There's one set of parentheses too many here.

`raise TraitError(object, name, 'a font descriptor string', repr( value ))`

**Comment by jdeschenes** [comment]
I fixed the issue you found and another that I saw from scanning the code.

File: enable/savage/svg/svg_regex.py [hunk 2]
```diff
@@ -114,9 +115,9 @@ def __init__(self, lexer=svg_lexer):
     def parse(self, text):
         """ Parse a string of SVG <path> data.
         """
-        next = self.lexer.lex(text).next
-        token = next()
-        return self.rule_svg_path(next, token)
+        svg_iterator = self.lexer.lex(text)
+        token = six.next(svg_iterator)
+        return self.rule_svg_path(svg_iterator, token)
```
Comment (jwiggins) [inline enable/savage/svg/svg_regex.py]: This appears to be incorrect. Perhaps something like: `return self.rule_svg_path(partial(six.next, svg_iterator), token)` or its equivalent. The first argument to `self.rule_svg_path` has to be a callable.

File: enable/savage/trait_defs/ui/toolkit.py [hunk 3]
```diff
@@ -31,9 +31,9 @@ def _init_toolkit():
     backend = 'enable.savage.trait_defs.ui.%s' % ETSConfig.toolkit
     try:
         __import__(backend)
-    except ImportError, SystemExit:
-        raise ImportError, "Unable to import a Savage backend for the %s " \
-            "toolkit." % ETSConfig.toolkit
+    except (ImportError, SystemExit):
```
Comment (jwiggins) [inline enable/savage/trait_defs/ui/toolkit.py]: Are the parentheses necessary?

File: enable/tests/container_test_case.py [hunk 4]
```diff
@@ -10,8 +12,8 @@ def assert_dims(self, obj, **dims):
         checks that each of the named dimensions of the object are a
         certain value.  e.g.   assert_dims(component, x=5.0, y=7.0).
         """
-        for dim, val in dims.items():
-            self.assert_( getattr(obj, dim) == val )
+        for dim, val in six.iteritems(dims):
```
Comment (jwiggins) [inline enable/tests/container_test_case.py]: Just use `items`

File: enable/tests/event_transform_test_case.py [hunk 5]
```diff
@@ -18,8 +20,8 @@ def assert_dims(self, obj, **dims):
         checks that each of the named dimensions of the object are a
         certain value.  e.g.   assert_dims(component, x=5.0, y=7.0).
         """
-        for dim, val in dims.items():
-            self.assert_( getattr(obj, dim) == val )
+        for dim, val in six.iteritems(dims.items):
```
Comment (jwiggins) [inline enable/tests/event_transform_test_case.py]: Just use `items`

File: enable/tests/primitives/test_image.py [hunk 6]
```diff
@@ -110,15 +113,17 @@ def test_image_gc_32(self):
         image_gc = self.image_32._image
         assert_array_equal(image_gc.bmp_array, self.data)
 
+    @unittest.skipIf(six.PY3, reason="Crashes on python 3. See GH #95.")
     def test_draw_24(self):
-        gc = GraphicsContext((256, 128), pix_format='rgba32')
+        gc = GraphicsContext((256, 128), pix_format='rgb24')
```
Comment (jwiggins) [inline enable/tests/primitives/test_image.py]: Why did you change `pix_format`?

File: enable/tools/value_drag_tool.py [hunk 7]
```diff
@@ -123,7 +124,7 @@ def _drag_button_down(self, event):
         # override button down to handle modifier keys correctly
         if not event.handled and self._drag_state == "nondrag":
             key_states = dict((key, key in self.modifier_keys) for key in keys)
-            if not all(getattr(event, key+'_down') == state for key, state in key_states.items()):
+            if not all(getattr(event, key+'_down') == state for key, state in six.iteritems(key_states)):
```
Comment (jwiggins) [inline enable/tools/value_drag_tool.py]: Just use `items`

File: enable/trait_defs/rgba_color_trait.py [hunk 8]
```diff
@@ -80,7 +82,7 @@ def convert_to_color ( object, name, value ):
 
 # RGBA versions of standard colors
 rgba_standard_colors = {}
-for name, color in standard_colors.items():
+for name, color in six.iteritems(standard_colors):
```
Comment (jwiggins) [inline enable/trait_defs/rgba_color_trait.py]: `items`

File: enable/vtk_backend/vtk_window.py [hunk 9]
```diff
@@ -281,6 +283,7 @@ def _vtk_mouse_wheel(self, vtk_obj, eventname):
             self._pass_event_to_vtk(vtk_obj, eventname)
 
     def _create_key_event(self, vtk_event, event_type):
+        # FIXME: THIS IS A BUG
```
Comment (jwiggins) [inline enable/vtk_backend/vtk_window.py]: What's the bug? It's not obvious from the diff here.

File: enable/vtk_backend/vtk_window.py [hunk 10]
```diff
@@ -290,11 +293,11 @@ def _create_key_event(self, vtk_event, event_type):
                 return self._pass_event_to_vtk(vtk_obj, eventname)
 
         if event_type == 'character':
-            key = unicode(self.control.key_sym)
+            key = six.u(self.control.key_sym)
```
Comment (jwiggins) [inline enable/vtk_backend/vtk_window.py]: Use `six.text_type` instead

File: enable/vtk_backend/vtk_window.py [hunk 11]
```diff
@@ -290,11 +293,11 @@ def _create_key_event(self, vtk_event, event_type):
...
         else:
             key = KEY_MAP.get(self.control.key_sym, None)
             if key is None:
-                key = unicode(self.control.key_sym)
+                key = six.u(self.control.key_sym)
```
Comment (jwiggins) [inline enable/vtk_backend/vtk_window.py]: Here too

File: enable/wx/cairo.py [hunk 12]
```diff
@@ -25,6 +27,7 @@ def _create_gc(self, size, pix_format="bgra32"):
 
     def _window_paint(self, event):
         "Do a GUI toolkit specific screen update"
+        # FIXME: THIS IS A BUG
```
Comment (jwiggins) [inline enable/wx/cairo.py]: What's the bug?

File: enable/wx/constants.py [hunk 13]
```diff
@@ -61,7 +64,7 @@
 if len(pointer_names) != len(pointer_shapes):
     warnings.warn("The WX toolkit backend pointer map is out of sync!")
 
-POINTER_MAP = dict(zip(pointer_names, pointer_shapes))
+POINTER_MAP = dict(sm.zip(pointer_names, pointer_shapes))
```
Comment (jwiggins) [inline enable/wx/constants.py]: Here and below, just leave the code as is.

File: kiva/agg/freetype2/src/tools/chktrcmp.py [hunk 14]
```diff
@@ -24,20 +25,20 @@
...
-    print "        Specify the directories of C source files to be checked"
-    print "        Default is %s" % ":".join( SRC_FILE_DIRS )
-    print ""
-    print "  --def-files=file1:file2:..."
-    print "        Specify the header files including FT_TRACE_DEF()"
-    print "        Default is %s" % ":".join( TRACE_DEF_FILES )
-    print ""
+    print("Usage: %s [option]" % sys.argv[0])
+    print("Search used-but-defined and defined-but-not-used trace_XXX macros")
+    print("")
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/chktrcmp.py]: You can just use `print()` wherever `print("")` is currently written

File: kiva/agg/freetype2/src/tools/docmaker/content.py [hunk 15]
```diff
@@ -439,14 +443,14 @@ def  finish( self ):
         # process all sections to extract their abstract, description
         # and ordered list of items
         #
-        for sec in self.sections.values():
+        for sec in six.itervalues(self.sections):
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/content.py]: Just use `values`

File: kiva/agg/freetype2/src/tools/docmaker/content.py [hunk 16]
```diff
@@ -459,7 +463,7 @@ def  finish( self ):
         # check that all sections are in a chapter
         #
         others = []
-        for sec in self.sections.values():
+        for sec in six.itervalues(self.sections):
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/content.py]: same here

File: kiva/agg/freetype2/src/tools/docmaker/tohtml.py [hunk 17]
```diff
@@ -169,29 +171,29 @@
...
 def  html_quote0( line ):
-    return string.replace( line, "&", "&amp;" )
+    return line.replace("&", "&amp;" )
 
 
 def  dump_html_code( lines, prefix = "" ):
     # clean the last empty lines
-    l = len( self.lines )
-    while l > 0 and string.strip( self.lines[l - 1] ) == "":
+    l = len( lines )
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/tohtml.py]: Was `self.lines` wrong?

File: kiva/agg/freetype2/src/tools/docmaker/tohtml.py [hunk 18]
```diff
@@ -491,22 +493,22 @@ def  toc_dump( self, toc_filename = None, index_filename = None ):
...
-        print section_title_header
-        print section.title
-        print section_title_footer
+        print(section_title_header)
+        print(section.title)
+        print(section_title_footer)
 
         maxwidth = 0
-        for b in section.blocks.values():
+        for b in six.itervalues(section.blocks):
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/tohtml.py]: Just use `values`

File: kiva/agg/freetype2/src/tools/docmaker/tohtml.py [hunk 19]
```diff
@@ -526,65 +528,65 @@ def  section_enter( self, section ):
...
         # place html anchor if needed
         if block.name:
-            print '<h4><a name="' + block.name + '">' + block.name + '</a></h4>'
+            print('<h4><a name="' + block.name + '">' + block.name + '</a></h4>')
 
         # dump the block C source lines now
         if block.code:
             header = ''
-            for f in self.headers.keys():
+            for f in six.iterkeys(self.headers):
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/tohtml.py]: just use `keys`

File: kiva/agg/freetype2/src/tools/glnames.py [hunk 20]
```diff
@@ -4860,7 +4860,8 @@ def add( self, word, value ):
 
   def optimize( self ):
     # optimize all children first
-    children      = self.children.values()
+    # Create a copy
+    children      = list(six.itervalues(self.children))
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/glnames.py]: `children = list(self.children.values())`

File: kiva/agg/freetype2/src/tools/glnames.py [hunk 21]
```diff
@@ -4908,8 +4909,7 @@ def locate( self, index ):
     if self.value != 0:
       index += 2
 
-    children = self.children.values()
-    children.sort()
+    children = sorted(six.itervalues(self.children))
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/glnames.py]: `children = sorted(self.children.values())`

File: kiva/agg/src/constants.i [hunk 22]
```diff
@@ -87,7 +88,7 @@ pix_format_string_map["abgr32"] = pix_format_abgr32
 pix_format_string_map["bgra32"] = pix_format_bgra32
 
 pix_format_enum_map = {}
-for key,value in pix_format_string_map.items():
+for key,value in six.iteritems(pix_format_string_map):
```
Comment (jwiggins) [inline kiva/agg/src/constants.i]: Just use `items` (here and below)

File: kiva/agg/src/font_type.i [hunk 23]
```diff
@@ -60,6 +60,9 @@ namespace kiva
 %{
 def unicode_safe_init(self, _name="Arial", _size=12, _family=0, _style=0,
                       _encoding=0, validate=True):
+    ### HACK:  C++ stuff expects a string (not unicode) for the face_name, so fix
+    ###        if needed.  See ticket #2111 in the CP Trac.
+    ### Only for python < 3
```
Comment (jwiggins) [inline kiva/agg/src/font_type.i]: CP Trac?

File: kiva/fonttools/font.py [hunk 24]
```diff
@@ -82,13 +84,13 @@ def __init__(self, face_name="", size=12, family=SWISS, weight=NORMAL,
...
             (type(weight) != type(NORMAL)) or (type(style) != type(NORMAL)) or \
-            (type(underline) != int) or (not isinstance(face_name, basestring)) or \
+            (type(underline) != int) or (not isinstance(face_name, six.string_types)) or \
             (type(encoding) != type(DEFAULT)):
                 raise RuntimeError("Bad value in Font() constructor.")
         ### HACK:  C++ stuff expects a string (not unicode) for the face_name, so fix
         ###        if needed.  See ticket #2111 in the CP Trac.
         ### Only for python < 3
-        if '' == b'' and isinstance(face_name, unicode):
+        if six.PY2 and isinstance(face_name, six.text_type):
```
Comment (jwiggins) [inline kiva/fonttools/font.py]: Ahhh... Now I see where the comment came from.

File: kiva/fonttools/font_manager.py [hunk 25]
```diff
@@ -471,7 +472,7 @@ def findSystemFonts(fontpaths=None, fontext='ttf'):
             else:
                 fontfiles[abs_path] = 1
 
-    return [fname for fname in fontfiles.keys() if os.path.exists(fname)]
+    return [fname for fname in six.iterkeys(fontfiles) if os.path.exists(fname)]
```
Comment (jwiggins) [inline kiva/fonttools/font_manager.py]: Just use `keys`

File: kiva/fonttools/font_manager.py [hunk 26]
```diff
@@ -523,7 +524,7 @@ def ttfFontProperty(fpath, font):
     *font* is a :class:`FT2Font` instance.
     """
     props = getPropDict(font)
-    name = props[(1, 0, 0, 1)]
+    name = props[(1, 0, 0, 1)].decode()
```
Comment (jwiggins) [inline kiva/fonttools/font_manager.py]: The `.decode()` makes sense in _both_ Python versions?

File: kiva/fonttools/font_manager.py [hunk 27]
```diff
@@ -993,12 +994,12 @@ def ttfdict_to_fnames(d):
...
-            for variantd in styled.values():
-                for weightd in variantd.values():
-                    for stretchd in weightd.values():
-                        for fname in stretchd.values():
+    for named in six.itervalues(d):
+        for styled in six.itervalues(named):
+            for variantd in six.itervalues(styled):
+                for weightd in six.itervalues(variantd):
+                    for stretchd in six.itervalues(weightd):
+                        for fname in six.itervalues(stretchd):
```
Comment (jwiggins) [inline kiva/fonttools/font_manager.py]: Just use `values`

File: kiva/pdfmetrics.py [hunk 28]
```diff
@@ -65,17 +71,17 @@ def parseAFMFile(afmFileName):
     needed for embedding.  A better parser would accept
     options for what data you wwanted, and preserve the
     order."""
-
-    lines = open(afmFileName, 'r').readlines()
+    with open(afmFileName, 'r') as f:
+        lines = [line.strip() for line in f]
+    #FIXME: This condition probably doesn't work...
```
Comment (jwiggins) [inline kiva/pdfmetrics.py]: How do you figure?

File: kiva/pdfmetrics.py [hunk 29]
```diff
@@ -65,17 +71,17 @@ def parseAFMFile(afmFileName):
     needed for embedding.  A better parser would accept
     options for what data you wwanted, and preserve the
     order."""
-
-    lines = open(afmFileName, 'r').readlines()
+    with open(afmFileName, 'r') as f:
+        lines = [line.strip() for line in f]
```
Comment (jwiggins) [inline kiva/pdfmetrics.py]: Isn't it problematic to strip the lines here if `'\r'` is being split on below?

File: kiva/pdfmetrics.py [hunk 30]
```diff
@@ -249,7 +254,7 @@ def freeze(self):
         self.frozen = 1
 
     def isEqual(self, other):
-        return ((enc.name == other.name) and (enc.vector == other.vector))
+        return ((self.name == other.name) and (self.vector == other.vector))
```
Comment (jwiggins) [inline kiva/pdfmetrics.py]: Was the usage of `enc` a `NameError`? Otherwise you're changing the semantics

File: kiva/pdfmetrics.py [hunk 31]
```diff
@@ -418,18 +423,17 @@ def _loadMetrics(self, afmFileName):
...
-        strBbox = topLevel.get('FontBBox', [0,0,1000,1000])
-        tokens = string.split(strBbox)
-        self.bbox = []
-        for tok in tokens:
-            self.bbox.append(string.atoi(tok))
+        try:
+            strBbox = topLevel['FontBBox']
+            self.bbox = [int(tok) for tok in strBbox.split()]
+        except KeyError:
+            self.bbox = [0,0,1000,1000]
```
Comment (jwiggins) [inline kiva/pdfmetrics.py]: It would be nice to maintain the structure here, but I'm not sure what it means to pass a list of ints to `string.split`

File: kiva/ps.py [hunk 32]
```diff
@@ -199,7 +210,7 @@ def device_draw_image(self, img, rect):
 
         Requires the Python Imaging Library (PIL).
         """
-        from kiva.compat import pilfromstring, piltostring
+        from kiva.compat import pilfromstring, piltostring, Image as PilImage
```
Comment (jwiggins) [inline kiva/ps.py]: Why the `Image as PilImage` import?

File: kiva/svg.py [hunk 33]
```diff
@@ -81,10 +88,6 @@ def default_filter(kw1):
     constants.JOIN_MITER: 'miter'
     }
 
-font_map = {'Arial': 'Helvetica',
-            }
-import _fontdata
-
```
Comment (jwiggins) [inline kiva/svg.py]: Why did you remove this?

File: kiva/tests/test_agg_drawing.py [hunk 34]
```diff
@@ -6,7 +6,7 @@
 class TestAggDrawing(DrawingImageTester, unittest.TestCase):
 
     def create_graphics_context(self, width, height):
-        return GraphicsContext((width, height))
+        return GraphicsContext((width, height), pix_format="rgba32")
```
Comment (jwiggins) [inline kiva/tests/test_agg_drawing.py]: What's the reason for the new keyword arg?

File: kiva/trait_defs/kiva_font_trait.py [hunk 35]
```diff
@@ -147,8 +147,8 @@ def validate ( self, object, name, value ):
         except:
             pass
 
-        raise TraitError, ( object, name, 'a font descriptor string',
-                            repr( value ) )
+        raise TraitError(( object, name, 'a font descriptor string',
+                            repr( value ) ))
```
Comment (jwiggins) [inline kiva/trait_defs/kiva_font_trait.py]: Too many parentheses here

File: setup.py [hunk 36]
```diff
@@ -289,9 +289,5 @@ def run(self):
           },
           platforms=["Windows", "Linux", "Mac OS-X", "Unix", "Solaris"],
           zip_safe=False,
-          use_2to3=True,
-          # The imports fixer makes breaking changes (replacing __builtin__
-          # with builtins) in the auto-generated SWIG files like
-          # kiva/agg/agg.py.
-          use_2to3_exclude_fixers=['lib2to3.fixes.fix_imports'],
+          use_2to3=False,
```
Comment (jwiggins) [inline setup.py]: Yay 🎉

**Comment by jwiggins** [comment]
In general: Thanks!!! 💯 This is important work and we really appreciate you taking it on (here and in the other ETS libraries).

I'm a bit swamped with other things right now, so I haven't had a chance to do anything more than read through these changes. Hopefully someone else can take a look...

File: kiva/pdfmetrics.py (hunk 29 above)
Comment (jdeschenes) [inline kiva/pdfmetrics.py]: Hmmm. this requires a refactor no matter what. Neither `string.split` nor `string.strip` exist in python 3

File: kiva/pdfmetrics.py (hunk 31 above)
Comment (jdeschenes) [inline kiva/pdfmetrics.py]: It just doesn't work in python 2.7. You would get a `AttributeError`. I don't know how you can keep the structure from before. This is the best I can come up with while respecting the intent.

File: kiva/pdfmetrics.py (hunk 30 above)
Comment (jdeschenes) [inline kiva/pdfmetrics.py]: The variable `enc` doesn't exist in this method. It was probably moved around and the value changed but someone didn't bother to change the variable name.

In general this particular module is quite problematic.

File: kiva/ps.py (hunk 32 above)
Comment (jdeschenes) [inline kiva/ps.py]: It's used in the function but never imported in the module.(Line 249). I can move the import at the top of the module, but that could have more important impact.

File: kiva/svg.py (hunk 33 above)
Comment (jdeschenes) [inline kiva/svg.py]: A few lines below, it is being imported. font_map is also defined a bit below.

File: kiva/fonttools/font_manager.py (hunk 26 above)
Comment (jdeschenes) [inline kiva/fonttools/font_manager.py]: In python 2, `decode()` is a no-op if the value is already a unicode, in python 3 it raises an AttributeError.

The getPropDict will generate a "dictionary" of bytes, that needs to be decoded at this point. This module is quite sketchy in terms of unicode vs bytes. It should probably be refactored in some ways. This is the best I could come up with minimal code change.

File: kiva/fonttools/font_manager.py [hunk 37]
```diff
@@ -673,7 +674,7 @@ def createFontList(fontfiles, fontext='ttf'):
     a list of TrueType fonts.  An AFM font list can optionally be
     created.
     """
-
+    # FIXME: This function is particularly difficult to debug
```
Comment (jdeschenes) [inline kiva/fonttools/font_manager.py]: I guess I should really highlight that this particular function caused me a lot of headaches. I think there are several issues in enable that stems from the error gobbling that this function does. #259 comes to mind, there are probably more.

File: kiva/fonttools/font.py (hunk 24 above)
Comment (jdeschenes) [inline kiva/fonttools/font.py]: I copied the comment, because I could not understand the meaning of the lines.

File: kiva/agg/freetype2/src/tools/chktrcmp.py (hunk 14 above)
Comment (jdeschenes) [inline kiva/agg/freetype2/src/tools/chktrcmp.py]: Done everywhere

File: enable/wx/constants.py (hunk 13 above)
Comment (jdeschenes) [inline enable/wx/constants.py]: Done.

File: enable/tests/primitives/test_image.py (hunk 6 above)
Comment (jdeschenes) [inline enable/tests/primitives/test_image.py]: The pix_format is probably wrong. This is a test_draw_24 and it is using rgba32. We might want to add another test to cover the normal case as well?

File: enable/qt4/base_window.py [hunk 38]
```diff
@@ -356,14 +358,14 @@ def _create_key_event(self, event_type, event):
                 return None
 
         if event_type == 'character':
-            key = unicode(event.text())
+            key = six.u(event.text())
```
Comment (jdeschenes) [inline enable/qt4/base_window.py]: Fixed

File: enable/qt4/constants.py [hunk 39]
```diff
@@ -13,6 +13,9 @@
 
 import warnings
 
+import six
```
Comment (jdeschenes) [inline enable/qt4/constants.py]: Fixed.

File: enable/savage/compliance/drawer.py [hunk 40]
```diff
@@ -43,7 +46,7 @@ def OnPaint(self, evt):
 class DrawFrame(wx.Frame):
     def __init__(self, parent, *args, **kwargs):
         wx.Frame.__init__(self, parent, *args, **kwargs)
-        self.pathOps = dict((k,v) for (k,v) in wx.GraphicsPath.__dict__.iteritems() if k.startswith("Add"))
+        self.pathOps = dict((k,v) for (k,v) in six.iteritems(wx.GraphicsPath.__dict__) if k.startswith("Add"))
```
Comment (jdeschenes) [inline enable/savage/compliance/drawer.py]: Fixed.

File: enable/savage/trait_defs/ui/toolkit.py (hunk 3 above)
Comment (jdeschenes) [inline enable/savage/trait_defs/ui/toolkit.py]: Yes, it will raise a `SyntaxError` otherwise

**Comment by jdeschenes** [comment]
Thanks for the quick review by the way!

File: enable/enable_traits.py [hunk 41]
```diff
@@ -75,7 +77,7 @@
     'dot':       array( [ 2.0, 2.0 ] ),
     'long dash': array( [ 9.0, 5.0 ] )
 }
-__line_style_trait_map_keys = __line_style_trait_values.keys()
+__line_style_trait_map_keys = list(six.iterkeys(__line_style_trait_values))
```
Comment (jwiggins) [inline enable/enable_traits.py]: It runs once when the module is imported, so I wouldn't worry about it

File: enable/savage/trait_defs/ui/toolkit.py (hunk 3 above)
Comment (jwiggins) [inline enable/savage/trait_defs/ui/toolkit.py]: Right. Can I blame exhaustion for asking silly questions? 😅

File: enable/tests/primitives/test_image.py (hunk 6 above)
Comment (jwiggins) [inline enable/tests/primitives/test_image.py]: Oh, I see. Good catch.

File: enable/vtk_backend/vtk_window.py (hunk 9 above)
Comment (jwiggins) [inline enable/vtk_backend/vtk_window.py]: Wow... Yeah, looks quite broken. Not sure when that happened.

File: enable/wx/cairo.py (hunk 12 above)
Comment (jwiggins) [inline enable/wx/cairo.py]: Ah... `union_bounds` is defined in `enable.base`

File: kiva/agg/freetype2/src/tools/docmaker/tohtml.py (hunk 17 above)
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/tohtml.py]: I would leave it. The `freetype2` included in kiva is used for its C code and nothing else.

File: kiva/agg/src/font_type.i (hunk 23 above)
Comment (jwiggins) [inline kiva/agg/src/font_type.i]: OK

File: kiva/fonttools/font_manager.py (hunk 26 above)
Comment (jwiggins) [inline kiva/fonttools/font_manager.py]: OK

File: kiva/pdfmetrics.py (hunk 29 above)
Comment (jwiggins) [inline kiva/pdfmetrics.py]: I agree, but you should probably be very careful about modifying the order in which things occur

File: kiva/pdfmetrics.py (hunk 30 above)
Comment (jwiggins) [inline kiva/pdfmetrics.py]: Yup, looks like a long forgotten move

File: kiva/pdfmetrics.py (hunk 31 above)
Comment (jwiggins) [inline kiva/pdfmetrics.py]: OK

File: kiva/ps.py (hunk 32 above)
Comment (jwiggins) [inline kiva/ps.py]: Ugh, yep. I see the `NameError` now... Carry on.

File: kiva/svg.py (hunk 33 above)
Comment (jwiggins) [inline kiva/svg.py]: That's just sloppy... Carry on.

File: kiva/pdfmetrics.py (hunk 28 above)
Comment (jwiggins) [inline kiva/pdfmetrics.py]: OK

File: enable/vtk_backend/vtk_window.py (hunk 9 above)
Comment (jdeschenes) [inline enable/vtk_backend/vtk_window.py]: Do you think we can just replace the variables with `vtk_event` and `event_type`? That looks right from here, but there is no test to check if that works.

File: kiva/agg/freetype2/src/tools/docmaker/tohtml.py (hunk 17 above)
Comment (jdeschenes) [inline kiva/agg/freetype2/src/tools/docmaker/tohtml.py]: Ok

File: enable/wx/cairo.py (hunk 12 above)
Comment (jdeschenes) [inline enable/wx/cairo.py]: Ok I changed it.

File: enable/tests/primitives/test_image.py (hunk 6 above)
Comment (jdeschenes) [inline enable/tests/primitives/test_image.py]: Is that the expected behavior?

**Comment by jdeschenes** [comment]
I have changed a few more things from @jwiggins comments. I think all comments except itervalues, iterkeys and iteritems have been answered. If not, please advise. I will make the last change tomorrow.

**Comment by jdeschenes** [comment]
I think I have resolved all issues. Let me know if there is anything else to correct.

File: enable/tests/primitives/test_image.py (hunk 6 above)
Comment (jwiggins) [inline enable/tests/primitives/test_image.py]: I think the original test code wasn't wrong. It's testing drawing a 24bit image to a 32-bit destination.

We weren't testing drawing to a 24bit destination before, so your changes don't need to be reverted.

File: enable/trait_defs/rgba_color_trait.py [hunk 42]
```diff
@@ -18,6 +18,8 @@
   RR is red, GG is green, and BB is blue.
 """
 
+import six
```
Comment (jwiggins) [inline enable/trait_defs/rgba_color_trait.py]: Unused import

File: enable/vtk_backend/vtk_window.py (hunk 9 above)
Comment (jwiggins) [inline enable/vtk_backend/vtk_window.py]: No, I don't.

`_create_key_event` is called from `AbstractWindow`, but the way the events are being routed from VTK here will throw an exception before `_create_key_event` can be called. The `FIXME` comment is fine for now.

File: kiva/agg/freetype2/src/tools/chktrcmp.py [hunk 43]
```diff
@@ -9,6 +9,7 @@
 import sys
 import os
 import re
+import six
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/chktrcmp.py]: Unused import

File: kiva/agg/freetype2/src/tools/docmaker/content.py [hunk 44]
```diff
@@ -5,9 +5,13 @@
 #  comment blocks and build more structured objects out of them.
 #
 
-from sources import *
-from utils import *
 import string, re
+import sys
+
+import six
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/content.py]: Unused imports (`sys` and `six`)

File: kiva/agg/freetype2/src/tools/docmaker/formatter.py [hunk 45]
```diff
@@ -1,9 +1,12 @@
 #  Formatter (c) 2002, 2004, 2007, 2008 David Turner <david@freetype.org>
 #
+import sys
 
-from sources import *
-from content import *
-from utils   import *
+import six
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/formatter.py]: Unused import

File: kiva/agg/freetype2/src/tools/docmaker/tohtml.py [hunk 46]
```diff
@@ -1,9 +1,11 @@
 #  ToHTML (c) 2002, 2003, 2005, 2006, 2007, 2008
 #    David Turner <david@freetype.org>
 
-from sources import *
-from content import *
-from formatter import *
+import six
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/tohtml.py]: `from __future__ import print_function` and you don't appear to be using `six` here

File: kiva/agg/freetype2/src/tools/glnames.py [hunk 47]
```diff
@@ -28,7 +28,7 @@
 
 
 import sys, string, struct, re, os.path
-
+import six
```
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/glnames.py]: Unused import

File: kiva/agg/src/constants.i [hunk 48]
```diff
@@ -53,6 +53,7 @@ unsigned path_flags(unsigned c);
 #   marker_enum_map[enum] = string
 #
 #----------------------------------------------------------------------------
+import six
```
Comment (jwiggins) [inline kiva/agg/src/constants.i]: Unused import

File: kiva/cairo.py [hunk 49]
```diff
@@ -1292,8 +1293,9 @@ def font_metrics_provider():
     from traits.api import false
     from chaco.api import ArrayPlotData, Plot, PlotGraphicsContext
     from chaco.example_support import COLOR_PALETTE
+    import six.moves as sm
```
Comment (jwiggins) [inline kiva/cairo.py]: This was already imported at the top of this module.

File: kiva/tests/test_pdfmetrics.py [hunk 50]
```diff
@@ -0,0 +1,55 @@
+import os
+import unittest
+import tempfile
+from contextlib import contextmanager
+
+from six.moves import StringIO
```
Comment (jwiggins) [inline kiva/tests/test_pdfmetrics.py]: This import appears to be unused

File: kiva/trait_defs/ui/wx/kiva_font_editor.py [hunk 51]
```diff
@@ -18,6 +18,7 @@
 #-------------------------------------------------------------------------------
 #  Imports:
 #-------------------------------------------------------------------------------
+import six
```
Comment (jwiggins) [inline kiva/trait_defs/ui/wx/kiva_font_editor.py]: Unused import

File: kiva/agg/freetype2/src/tools/docmaker/content.py (hunk 44 above)
Comment (jdeschenes) [inline kiva/agg/freetype2/src/tools/docmaker/content.py]: `sys` is used in the module but was never imported.

**Comment by jdeschenes** [comment]
Fixed that unused imports. There was another issue that went unnoticed in vu_meter that I fixed.

File: kiva/agg/freetype2/src/tools/docmaker/content.py (hunk 44 above)
Comment (jwiggins) [inline kiva/agg/freetype2/src/tools/docmaker/content.py]: Good catch

**Comment by jwiggins** [comment]
I think I'm basically done reviewing this. Does anybody with fresh eyes want to take a look before it's merged?

File: kiva/quartz/ABCGI.pyx [hunk 52]
```diff
@@ -2619,28 +2619,28 @@ cdef class PiecewiseLinearColorFunction(ShadingFunction):
...
+        print('PiecewiseLinearColorFunction')
+        print('  num_stops = %i' % self.num_stops)
+        print('  stops = ', end=" ")
         for i from 0 <= i < self.num_stops:
-            print self.stops[i],
-        print
-        print '  red = ',
+            print(self.stops[i], end=" ")
+        print()
+        print('  red = ', end=" "
```
Comment (jwiggins) [inline kiva/quartz/ABCGI.pyx]: This line is missing its closing `)`

**Comment by jwiggins** [comment]
OK, so I finally had a chance to test this. Looks good, modulo the missing parenthesis which causes a build failure on macOS.

**Comment by jdeschenes** [comment]
I added the parenthesis. Building MacOS on travis should be done quite quickly so this doesn't slip through!

**Comment by jwiggins** [comment]
Thanks for all the updates @jdeschenes!
